app/src/behaviors/behavior_tap_dance.c

```diff
@@ -0,0 +1,266 @@
+    };
+    behavior_keymap_binding_pressed(&binding, event);
+    return 0;
+}
+
+static inline int release_tap_dance_behavior(struct active_tap_dance *tap_dance,
+                                             int64_t timestamp) {
+    LOG_DBG("Release Tap Dance Behavior");
+    struct zmk_behavior_binding binding =
+        (tap_dance->counter < tap_dance->config->behavior_count)
+            ? tap_dance->config->behaviors[(tap_dance->counter) - 1]
+            : tap_dance->config->behaviors[(tap_dance->config->behavior_count) - 1];
+    struct zmk_behavior_binding_event event = {
+        .position = tap_dance->position,
+        .timestamp = timestamp,
+    };
+    behavior_keymap_binding_released(&binding, event);
+    clear_tap_dance(tap_dance);
+    return 0;
+}
+
+static int on_tap_dance_binding_pressed(struct zmk_behavior_binding *binding,
+                                        struct zmk_behavior_binding_event event) {
+    LOG_DBG("On Binding Pressed");
+    const struct device *dev = device_get_binding(binding->behavior_dev);
+    const struct behavior_tap_dance_config *cfg = dev->config;
+    struct active_tap_dance *tap_dance;
+    tap_dance = find_tap_dance(event.position);
+    if (tap_dance == NULL) {
+        tap_dance = store_tap_dance(event.position, cfg);
+        reset_timer(tap_dance, event);
+        return ZMK_BEHAVIOR_OPAQUE;
+    }
+    tap_dance->is_pressed = true;
+    stop_timer(tap_dance);
+    if (++tap_dance->counter >= cfg->behavior_count) {
+        press_tap_dance_behavior(tap_dance, event.timestamp);
+        tap_dance->tap_dance_decided = true;
+        return ZMK_BEHAVIOR_OPAQUE;
+    }
+    reset_timer(tap_dance, event);
+    return ZMK_BEHAVIOR_OPAQUE;
+}
+
+static int on_tap_dance_binding_released(struct zmk_behavior_binding *binding,
+                                         struct zmk_behavior_binding_event event) {
+    LOG_DBG("On Binding Released");
+    struct active_tap_dance *tap_dance = find_tap_dance(event.position);
+    if (tap_dance == NULL) {
+        LOG_ERR("ACTIVE TAP DANCE CLEARED TOO EARLY");
+        return ZMK_BEHAVIOR_OPAQUE;
+    }
+    tap_dance->is_pressed = false;
+    if (tap_dance->tap_dance_decided) {
+        release_tap_dance_behavior(tap_dance, event.timestamp);
+        tap_dance->tap_dance_decided = false;
+    }
+    return ZMK_BEHAVIOR_OPAQUE;
+}
+
+void behavior_tap_dance_timer_handler(struct k_work *item) {
+    LOG_DBG("Timer Handler Called");
+    struct active_tap_dance *tap_dance = CONTAINER_OF(item, struct active_tap_dance, release_timer);
+    if (tap_dance->position == ZMK_BHV_TAP_DANCE_POSITION_FREE) {
+        return;
+    }
+    tap_dance->timer_cancelled = tap_dance->timer_cancelled ? false;
+    if (tap_dance->is_pressed) {
+        press_tap_dance_behavior(tap_dance, tap_dance->release_at);
+        tap_dance->tap_dance_decided = true;
+        return;
+    }
+    press_tap_dance_behavior(tap_dance, tap_dance->release_at);
```

Comment:
```     press_tap_dance_behavior(tap_dance, tap_dance->release_at);```
this code is called in the if and after it, so it can be done before the if statement.